### PR TITLE
Calling dockerized ort from TypeScript

### PR DESCRIPTION
**Pull request by jspaaks**
## Description

Refs: #34, #45

I this PR, I'm experimenting with calling dockerized ort from TypeScript. The basics are there, but needs more work.

* * *

## Before you get started

- [ ] [☝ Create an issue](https://github.com/NLeSC/licenseguard/issues/new/choose) to discuss what you are going to do

## Checklist

This section should be filled in by the creator of this pull request to make sure the pull request is ready to review.

- [x] This pull request has a descriptive title
- [x] Code is written according to the code quality guidelines
- [x] Documentation if applicable is available
- [x] Tests run successfully
- [x] All checks below this pull request were successful

## Instructions to review the pull request

```shell
git clone this repo
cd licenseguard
git checkout this branch
npm install
npm run build
npm run package
MYDIR=`pwd`

# cd to tempdir to facilitate writing the output and downloading the clones
cd $(mktemp -d --tmpdir ortolan.XXXXXX)
node $MYDIR/dist/index.js
```

Kinda works (see result below) but there's something odd with the async 

```shell
.
├── in
│   ├── iomega
│   │   └── zenodo-upload
│   │       ├── CHANGELOG.md
│   │       ├── CITATION.cff
│   │       ├── CODE_OF_CONDUCT.md
│   │       ├── CONTRIBUTING.md
│   │       ├── docs
│   │       ├── LICENSE
│   │       ├── package.json
│   │       ├── README.md
│   │       ├── sonar-project.properties
│   │       ├── src
│   │       ├── test
│   │       ├── tsconfig.json
│   │       └── yarn.lock
│   └── xenon-middleware
│       └── xenon-cli
│           ├── appveyor.yml
│           ├── build.gradle
│           ├── CHANGELOG.md
│           ├── CITATION.cff
│           ├── conda
│           ├── Dockerfile
│           ├── gradle
│           ├── gradlew
│           ├── gradlew.bat
│           ├── LICENSE
│           ├── README.md
│           ├── src
│           ├── xenon-download.cwl
│           ├── xenon-ls.cwl
│           └── xenon-upload.cwl
└── out
    ├── iomega
    │   └── zenodo-upload
    │       └── analyzer-result.yml
    └── xenon-middleware
        └── xenon-cli
            └── analyzer-result.yml
```

```
$ cat out/iomega/zenodo-upload/analyzer-result.yml | head -n 10
---
repository:
  vcs:
    type: "Git"
    url: "https://github.com/iomega/zenodo-upload"
    revision: "4b850962abc1f467071d9e09348cb67deb4c8bb7"
    path: ""
  vcs_processed:
    type: "Git"
    url: "https://github.com/iomega/zenodo-upload.git"

```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,13 +22,14 @@
...
   ],
-  "author": "",
+  "author": "Netherlands eScience Center",
   "license": "Apache-2.0",
   "dependencies": {
     "@actions/core": "^1.2.6",
     "js-yaml": "^4.1.0"
   },
   "devDependencies": {
+    "@actions/exec": "^1.0.4",
```
Comment (sverhoeven) [inline package.json]: Please move to dependencies as exec is used on runtime, not during development (`npm run *`).

Comment (jspaaks) [reply package.json]: yep

**Comment by jspaaks** [comment]
Marked as _Ready for review_ because this could use some input from others

File: .github/workflows/linting.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+name: 'linting'
```
Comment (sverhoeven) [inline .github/workflows/linting.yml:1]: Why did you split the test workflow into multiple workflows?

File: src/ortolan.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import * as shell from '@actions/exec'
```
Comment (sverhoeven) [inline src/ortolan.ts]: Simplify import by using named import instead of wildcard.
```suggestion
import {exec) from '@actions/exec'
```
When suggestion is commited the `shell.exec` occurences should be replaced with `exec`.

File: tsconfig.json [hunk 4]
```diff
@@ -6,7 +6,8 @@
     "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
     "strict": true,                           /* Enable all strict type-checking options. */
     "noImplicitAny": true,                    /* Raise error on expressions and declarations with an implied 'any' type. */
-    "esModuleInterop": true                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "esModuleInterop": true,                  /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "sourceMap": true
```
Comment (sverhoeven) [inline tsconfig.json:10]: This was probably added to debug running `dist/index.js` with breakpoints in src/*. Would be nice to see how you can do this.

File: .github/workflows/testing.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      - uses: actions/checkout@v2
+      - run: |
+          npm install
+      - run: |
+          npm run build
+      - run: |
+          npm run package
+      - run: |
+          npm run test
+
```
Comment (sverhoeven) [inline .github/workflows/testing.yml]: Thie original test.yml also ran the action on `.` why was this deleted?

**Review by sverhoeven — CHANGES_REQUESTED** [review]
This PR does more than just running ort analyze. So I have some questions about those other bits.

File: tsconfig.json (hunk 4 above)
Comment (sverhoeven) [reply tsconfig.json:10]: I normally to debugging by starting a unit test in debug mode. Would be nice to see how to debugging on a package file.

**Comment by jspaaks** [comment]
This code is using ort, but that needs to be built first from its repo; maybe update the name to philipssoftware/ort

File: .github/workflows/linting.yml (hunk 2 above)
Comment (jspaaks) [reply .github/workflows/linting.yml:1]: They are conceptually different, so I'd like to see different status badges for each

File: src/ortolan.ts (hunk 3 above)
Comment (jspaaks) [reply src/ortolan.ts]: done in upcoming commit

File: .github/workflows/testing.yml (hunk 5 above)
Comment (jspaaks) [reply .github/workflows/testing.yml]: reinstated the `.` without the `wait` since that doesnt exist anymore

File: src/ortolan.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    constructor(repositories: Repositories) {
+        this.repositories = repositories
+    }
+
+    analyze(owner: string, repo: string): void {
+        const cwd = process.cwd()
+        const bindMountInput = `${cwd}/in/${owner}/${repo}:/project`
+        const bindMountOutput = `${cwd}/out/${owner}/${repo}:/out`
+        exec(
+            `docker run --rm -v ${bindMountInput} -v ${bindMountOutput} ort analyze -i /project -o /out`
```
Comment (sverhoeven) [inline src/ortolan.ts]: Make CI green by giving publicly available image name
```suggestion
        const bindMountOutput = `${cwd}/out/${owner}/${repo}:/out`
        const image = 'philipssoftware/ort'
        exec(
            `docker run --rm -v ${bindMountInput} -v ${bindMountOutput} ${image} analyze -i /project -o /out`
```

**Review by sverhoeven — APPROVED** [review]
Doh to make test-action run OK we should also update dist files.

test-action job does not fail, but the current code does not seem to cause the non-zero exit code to be set. Code should have [try/catch with `core.setFailed('Doh')`](https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions#setting-a-failure-exit-code-in-a-javascript-action) to have proper exit code.

However now the PR is better and did not loose any functionality.

**Comment by jspaaks** [comment]
Thanks for reviewing. I'll merge it and then we can work on fixing some of the shortcomings we discovered
